src/librustc_typeck/check/_match.rs

```diff
@@ -112,14 +115,74 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 self.check_expr_with_expectation(&arm.body, expected)
             };
             all_arms_diverge &= self.diverges.get();
+
+            // When we have a `match` as a tail expression in a `fn` with a returned `impl Trait`
+            // we check if the different arms would work with boxed trait objects instead and
+            // provide a structured suggestion in that case.
+            let suggest_box = match (
+                orig_expected,
+                self.body_id
+                    .owner
+                    .to_def_id()
+                    .as_local()
```

Comment:
`HirId -> LocalDefId -> DefId -> Option<LocalDefId>`.

This seems incorrect to me, is there a reason you don't use `tcx.hir().opt_local_def_id(self.body_id)`?

Not really part of the PR, but why is `body_id` a `HirId` instead of a `BodyId` for `FnCtxt`?

Reply:
Using the proposed change causes us to lose the following suggestion:

```
error[E0308]: `if` and `else` have incompatible types
 --> fil3.rs:5:9
  |
2 | /     if false {
3 | |         0i32
  | |         ---- expected because of this
4 | |     } else {
5 | |         1u32 //~ ERROR `if` and `else` have incompatible types
  | |         ^^^^ expected `i32`, found `u32`
6 | |     }
  | |_____- `if` and `else` have incompatible types
  |
help: you could change the return type to be a boxed trait object
  |
1 | fn qux() -> Box<dyn std::fmt::Display> {
  |             ^^^^^^^                  ^
help: if you change the return type to expect trait objects, box the returned expressions
  |
3 |         Box::new(0i32)
4 |     } else {
5 |         Box::new(1u32) //~ ERROR `if` and `else` have incompatible types
  |
```

I think that the following is equivalent to what we had before:

```rust
            let body_parent_id = tcx.hir().get_parent_did(self.body_id);
            let opt_suggest_box_span = match (
                orig_expected,
                self.ret_coercion_impl_trait.map(|ty| (body_parent_id, ty))
```

Reply:
Nevermind, just realized what a mess I got myself into that could be simplified quite a bit.